CMSgov/mint-app
# Easi 2572/data loaders spike

# EASI-2572

## Changes and Description

- Implement Initial Data Loaders
    - Plan Basics
    - Operational Needs
    - Operational Solutions
- [Data Loader ADR](https://github.com/CMSgov/mint-app/blob/EASI-2572/data_loaders_spike/docs/adr/0009-data-loaders.md)



## How to test this change
1. Seed Data
2. Make multiple model plans
3.  Run this query which will test all the new data loaders. If you like, you can enable debugging to verify that the database is only running 1 optimized query for each section (basics, operationalNeeds, operationalSolutions). Alternatively, you can look at the docker database logs to verify.
```gql
query modelPlanCollection {
    modelPlanCollection {
        id
        modelName
        basics{
            modelPlanID        
            id
            modelCategory
        }
        operationalNeeds {
 
            id
            modelPlanID
            name
            key
            # section
            nameOther
            needed
            solutions(includeNotNeeded:true) {
                # id
                # status
                # mustStartDts
                
                
                name
                key   
                needed
                nameOther
                pocEmail
                pocName     
                createdBy
                createdDts
            }
        }
        
    }
}
```

As this is a spike, there are still implementation details that should be added later, but the code is functional as an initial implementation. Unit tests have a basics implementation, but it would be great to more fully test them. Concurrency of is being tested using `errgroups`, as initial testing of assertions with a go routine revealed that is not a thread safe operation.

## PR Author Review Checklist



## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.
